### ISSUE TITLE
[Serializer] Custom Normalizer broken after upgrading to 6.1

### ISSUE THREAD

**Issue by alexchuin** [symfony/symfony#46471]
### Symfony version(s) affected

6.1

### Description

My custom Normalizer is broken since i've upgraded to Symfony 6.1.
This problem seems to be related to the new Serializer Profiler.

Maybe I forgot something?

Thanks,
Alex

### How to reproduce

Just use this example :

https://symfony.com/doc/6.1/serializer/custom_normalizer.html

### Additional Context

`FileNormalizer::__construct(): Argument #2 ($normalizer) must be of type Symfony\\Component\\Serializer\\Normalizer\\ObjectNormalizer, Symfony\\Component\\Serializer\\Debug\\TraceableNormalizer given`

**Comment by xabbuh** [comment]
Does the issue go away when you change the type-hint of your constructor argument from `ObjectNormalizer` to `NormalizerInterface` and does your normalizer still work as intended?

**Comment by alexchuin** [comment]
Nope, already tried and got a Circular reference detected :

`Circular reference detected for service "debug.App\Api\Normalizer\FileNorma  
!!    lizer", path: "debug.App\Api\Normalizer\FileNormalizer -> debug.serializer   
!!    -> debug.serializer.inner -> debug.App\Api\Normalizer\FileNormalizer". `

**Comment by xabbuh** [comment]
I changed my suggestion for the update of the documentation a bit. Can you try out the new version in https://github.com/symfony/symfony-docs/pull/16826 please?

**Comment by derrabus** [comment]
I think that the example from the docs is correct and should continue to work after an upgrade. We should treat this as a bug.

The serializer topic aside: If I autowire a constructor parameter with a `Foo` type, I find it astonishing that I receive a service that is not a `Foo`.

**Comment by alexchuin** [comment]
Yes, thanks all but I just downgraded to 6.0 until it will be fixed :)

**Comment by darthf1** [comment]
Similar (?) issue. On 6.0, the following service declaration works:
`services.yaml`
```yaml
  Infrastructure\Shared\ApiPlatform\Serializer\JsonLdNormalizer:
    decorates: 'api_platform.jsonld.normalizer.item'
```

```php
use ApiPlatform\Core\Api\IriConverterInterface;
use ApiPlatform\Core\JsonLd\Serializer\ItemNormalizer;
use Symfony\Component\Serializer\Normalizer\CacheableSupportsMethodInterface;
use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
use Symfony\Component\Serializer\Normalizer\NormalizerInterface;

final class JsonLdNormalizer implements NormalizerInterface, DenormalizerInterface, SerializerAwareInterface, CacheableSupportsMethodInterface
{
    public function __construct(
        private readonly IriConverterInterface $iriConverter,
        private readonly ItemNormalizer $decorated,
    ) {
    }

    ......
}
```

On 6.1 I suddenly get the error:
```
Cannot autowire service "Infrastructure\Shared\ApiPlatform\Serializer\JsonLdNormalizer": argument "$decorated" of method "__construct()" references class "ApiPlatform\Core\JsonLd\Serializer\ItemNormalizer" but no such   
  service exists. Try changing the type-hint to one of its parents: interface "Symfony\Component\Serializer\Normalizer\DenormalizerInterface", or interface "Symfony\Component\Serializer\Normalizer\NormalizerInterface". 
```

If I change my service declaration:
```yaml
  Infrastructure\Shared\ApiPlatform\Serializer\JsonLdNormalizer:
    decorates: 'api_platform.jsonld.normalizer.item'
    arguments:
      $decorated: '@.inner'
```

The error message changes:

```
Infrastructure\Shared\ApiPlatform\Serializer\JsonLdNormalizer::__construct(): Argument #2 ($decorated) must be of type ApiPlatform\Core\JsonLd\Serializer\ItemNormalizer,  
     Symfony\Component\Serializer\Debug\TraceableNormalizer given
```

**Comment by camilleterol** [comment]
I am having the same issue. Let me know if you need any help reproducing this one.

It happens when a Custom Normalizer is autowired with the ObjectNormalizer service.

**Comment by stof** [comment]
Well, I'm not sure the normalizer services were actually meant to be autowirable based on their class name. Supporting that means that we don't support decorating them (and so that we cannot apply the traceable decorators on them).

To me, the autowirable names in core should only about interfaces for that reason.

**Comment by derrabus** [comment]
I think, the use case described in the linked piece of documentation is a valid one: We want a normalizer that behaves like `ObjectNormalizer` mostly, but wee need to change a thing or two in the resulting array.

This is why our custom normalizer requests exactly that implementation of a normalizer. And our custom normalizer may receive an undecorated `ObjectNormalizer`, that's fine. I doubt that we need that level of traceability.

**Comment by danielburger1337** [comment]
A possible workaround, that would work with the new TraceableNormalizer, is to use the new `Autowire` attribute:

```php
class CustomNormalizer implements NormalizerInterface, CacheableSupportsMethodInterface
{
    public function __construct(
        #[Autowire(service: ObjectNormalizer::class)]
        private NormalizerInterface $normalizer
    ) {
    }
}
```

But I agree with @derrabus that it is somewhat crazy, that symfony injects an object with an incompatiable type. 

I'm not to familiar with how symfonys dependency injection works internally, but is the "symfony/proxy-manager-bridge" package not able to solve this problem?  A "lazy" service will be replaced by a proxy service, so wouldn't it be possible to create a proxy object that has the TraceableNormalizer functionality? Currently the service definition gets replaced by a decorator.

The library the bridge is based on, [Ocarmius/ProxyManager](https://github.com/Ocramius/ProxyManager/), has the [Access Interceptor Value Holder Proxy](https://github.com/Ocramius/ProxyManager/blob/2.15.x/docs/access-interceptor-value-holder.md) concept. This should be able to solve the problem.
Bascially I'm suggesting that the service is not replaced by "TraceableNormalizer" (this applies for all traceable decorators, like TraceableAuthenticator), but by a proxy object.

**Comment by aniolekx** [comment]
this "possible workaround" does not work for me :/ (Circular reference detected)

**Comment by shouze** [comment]
@danielburger1337 @aniolekx I confirm that workaround with Autowire works perfectly here, thx!

**Comment by shouze** [comment]
**EDIT**: still have the issue in local dev with profiler enabled in fact...

**Comment by shouze** [comment]
Yes I import the class, but I have another issue after. If you decorate a `Normalizer` that strictly implement the Normalize interface it's OK. But for Example we decorate the pagination Normalizer of Api Platform... that have some abstract methods like `getPaginationData`... and then in local dev I have a crash saying that TraceableNormalizer don't have any `getPaginationData` method :sob:

**Comment by danielburger1337** [comment]
> But for Example we decorate the pagination Normalizer of Api Platform... that have some abstract methods like `getPaginationData`... and then in local dev I have a crash saying that TraceableNormalizer don't have any `getPaginationData` method 😭

That is a good point. This could be easily fixed by adding php's magic `__call` method to the TraceableNormalizer implementation. As far as I can see, exactly this was already implemented for TreaceableAuthenticator, TraceableEventDispatcher, LoggingTranslator etc..

I can submit a PR incase this function wasn't omitted on purpose?

**Comment by chalasr** [comment]
PR welcome for the missing `__call()` method.

**Comment by chalasr** [comment]
> I think, the use case described in the linked piece of documentation is a valid one

While the use case is valid, the way it is implemented seems wrong to me given how the serializer component is designed, as stof explained: `NormalizerInterface`' concrete implementations are not meant to be referenced by something else than `Serializer` itself. To me, either it should be extended using inheritance, or we are missing some proper extension point here.

That doesn't mean we shouldn't try to fix this regression though.

**Comment by derrabus** [comment]
> `NormalizerInterface` concrete implementations are not meant to be referenced by something else than `Serializer` itself.

Not sure about that. In the documented case, the developer does not just want any normalizer implementation. They explicitly want the behavior of `ObjectNormalizer` because it solves 95% of their problem already. The alternatives are either to extend `ObjectNormalizer` or to work with `NormalizerAwareInterface` and work with priorities and flags that are pushed to the context.

The former is a technique which we should discourage imho. And the latter feels a bit too complicated for what should be achieved here.

**Comment by silverbackdan** [comment]
Could we address @darthf1 use case as this is mine as well - decorating a specific normalizer will result in this issue. We should be able to decorate like any other service IMHO. So we can create a new service, and type hint the specific normalizer we are wanting to be passed into the constructor.

**Comment by chalasr** [comment]
>  the developer does not just want any normalizer implementation. They explicitly want the behavior of ObjectNormalizer because it solves 95% of their problem already.

My point is that this should be done via an explicit service definition so the code only knows about NormalizerInterface, and the config tells which implementation should actually be injected. Targeting ObjectNormalizer directly as a constructor arg should be avoided from a design POV. Same for the "decoration" use case, proper decoration is made against an abstraction, not a concrete implementation.

**Comment by Seldaek** [comment]
Sorry if this is a dumb question, but can we not fix this in the DIC, making sure that if a decorated service is selected to be autowired, but the decorator does not satisfy the type hint, it would skip decoration and inject the decorated service?

Possibly this could be done optionally so that decorators could be marked optional (as in this case, it's a "nice to have" feature, a user defined decorator I would assume should error if it cannot be autowired as such).

**Comment by ahmetilhann** [comment]
I think this feature should be reverted. Nested normalizers are an overused method.
Or the old and new should continue to work at the same time.
I will continue with 6.0 for a while.

**Comment by chalasr** [comment]
> Sorry if this is a dumb question, but can we not fix this in the DIC, making sure that if a decorated service is selected to be autowired, but the decorator does not satisfy the type hint, it would skip decoration and inject the decorated service?

@Seldaek Not a dumb question for sure, but I personally don't see where/how that skipping logic could be implemented. 
Here we are basically decorating all normalizers from a compiler pass that doesn't know where these decorated services are referenced (nor if they are explicitly or via autowiring, both are broken for the use case at hand). 
What I'm pretty sure is that looking into all those referencing services to check if their type declaration would fit with the decorator type would have serious performance implications. Tackling this in a generic way by allowing to mark a decorator as optional would probably be possible, at least worth trying, but that would be a new feature requiring some iterations at least to make sure it doesn't cause perf issues.

> I think this feature should be reverted. Nested normalizers are an overused method.

Fact is that it is not inline with the component's current design and that it forbids us to decorate normalizers in any way. IMHO, it's a sign we would need to discourage such usage in favor of a better alternative (which exists but is not documented, see e.g. https://symfonycasts.com/screencast/api-platform-security/normalizer-aware).
And yes, ideally we should have anticipated this, but we did not.

**Comment by chalasr** [comment]
See #46625

**Comment by stof** [comment]
@zim32 decorating normalizers is off by default in the minor release (this was not in a patch release at all). If you don't modify your config to enable it, nothing breaks.

**Comment by zim32** [comment]
> @zim32 decorating normalizers is off by default in the minor release (this was not in a patch release at all). If you don't modify your config to enable it, nothing breaks.

I was extenging ObjectNormalizer. So it breaks my code. Ok I decide to rewrite using composition but now have problem described in this issue. 

So how to implemnet custom normalizer/denormalizer nowadays? Should I create factory and construct serializer from scratch? Which method to use?

**Comment by zim32** [comment]
There was method with NormalizerAwareInterface? where I can delegate to inner normalizer passing some param to context, and then check this param in supports... method to prevent intinite loop. By they removed context from supports method. Now I can not check this.

**Comment by zim32** [comment]
> @zim32 from [symfony/symfony-docs#16826](https://github.com/symfony/symfony-docs/pull/16826):
> 
> > Change the parameter type declaration from ObjectNormalizer to NormalizerInterface and add the config needed to make it gets the serializer.normalizer.object service injected
> 
> (i.e. define your service and give it the right normalizer as argument explicitly)

See previous message about recursion and context

**Comment by mtarld** [comment]
You still can access the context in the NormalizerInterface:

```php
public function supportsNormalization(mixed $data, string $format = null /* , array $context = [] */);
```

It behaves exactly as the NormalizerAwareInterface did.

**Comment by zim32** [comment]
> You still can access the context in the NormalizerInterface:
> 
> ```
> public function supportsNormalization(mixed $data, string $format = null /* , array $context = [] */);
> ```
> 
> It behaves exactly as the NormalizerAwareInterface did.

How? If add add array $context, my supports method will incompatible with interface

**Comment by zim32** [comment]
Ok the only way I found to inject denormalizer into my own denormalizer is to disable autowiring and inject **@serializer.normalizer.object** directly.

**Comment by chalasr** [comment]
> If I add array $context, my supports method will be incompatible with interface

Nope, it won't: https://3v4l.org/sKOnn

> Ok the only way I found to inject denormalizer into my own denormalizer is to disable autowiring and inject @serializer.normalizer.object directly.

That's my personal preference.

**Comment by zim32** [comment]
> > If I add array $context, my supports method will be incompatible with interface
> 
> Nope, it won't: https://3v4l.org/sKOnn
> 
> > Ok the only way I found to inject denormalizer into my own denormalizer is to disable autowiring and inject @serializer.normalizer.object directly.
> 
> That's my personal preference.

Ah really. I forget to add  = [], as default parameter.

**Comment by lyrixx** [comment]
I'm a bit late on this one, but the bug is still here and it's easy to reproduce **without custom code**


```
symfony new --webapp  /tmp/serializer-ko
cd /tmp/serializer-ko
bin/console make:controller Main
# Inject the serializer into the controller
# Open the page => OK
bin/console make:serializer:normalizer Foobar
# Open the page => KO
```



> App\Serializer\Normalizer\UserNormalizer::__construct(): Argument #1 ($normalizer) must be of type Symfony\Component\Serializer\Normalizer\ObjectNormalizer, Symfony\Component\Serializer\Debug\TraceableNormalizer given, called in /tmp/debug/serializer-ko/var/cache/dev/Container04xllCG/getDebug_SerializerService.php on line 80
